src/dependencies.rs

```diff
@@ -0,0 +1,305 @@
+        };
+
+        if !string.contains("==") {
+            return Err(err);
+        }
+
+        let words: Vec<_> = string.split("==").collect();
+        if words.len() != 2 {
+            return Err(err);
+        }
+        Ok(FrozenDependency {
+            name: words[0].to_string(),
+            version: words[1].to_string(),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct ParseError {
+    pub details: String,
+}
+
+impl ParseError {
+    pub fn new(details: &str) -> Self {
+        ParseError {
+            details: details.to_string(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct VersionSpec {
+    start: usize,
+    end: usize,
+    pub value: String,
+}
+
+#[derive(Debug)]
+pub struct GitDependency {
+    pub name: String,
+    pub line: String,
+    pub git_ref: VersionSpec,
+}
+
+impl GitDependency {
+    pub fn bump(&mut self, new_ref: &str) -> bool {
+        let VersionSpec { start, end, value } = &self.git_ref;
+        if new_ref == value {
+            return false;
+        }
+        self.line = format!("{}{}{}", &self.line[0..*start], new_ref, &self.line[*end..],);
+        self.git_ref.value = new_ref.to_string();
+        true
+    }
+}
+
+#[derive(Debug)]
+pub struct SimpleDependency {
+    pub name: String,
+    pub version: VersionSpec,
+    pub line: String,
+}
+
+impl SimpleDependency {
+    pub fn from_frozen(frozen: &FrozenDependency) -> Self {
+        let name = &frozen.name;
+        let line = format!("{}=={}", name, frozen.version);
+        let version = LockedDependency::parse_simple_version(&line);
+        SimpleDependency {
+            name: name.to_string(),
+            version,
+            line,
+        }
+    }
+
+    pub fn bump(&mut self, new_version: &str) -> bool {
+        let VersionSpec { start, end, value } = &self.version;
+        if new_version == value {
+            return false;
+        }
+        self.line = format!(
+            "{}{}{}",
+            &self.line[0..*start],
+            new_version,
+            &self.line[*end..],
+        );
+        self.version.value = new_version.to_string();
+        true
+    }
+
+    pub fn freeze(&mut self, new_version: &str) {
+        self.bump(new_version);
+    }
+}
+
+#[derive(Debug)]
+pub enum LockedDependency {
+    Git(GitDependency),
+    Simple(SimpleDependency),
+}
+
+impl LockedDependency {
+    pub fn line(&self) -> String {
+        match self {
+            LockedDependency::Git(x) => x.line.to_string(),
+            LockedDependency::Simple(x) => x.line.to_string(),
+        }
+    }
+
+    pub fn name(&self) -> String {
+        match self {
+            LockedDependency::Git(x) => x.name.to_string(),
+            LockedDependency::Simple(x) => x.name.to_string(),
+        }
+    }
+
+    pub fn from_line(line: &str) -> Result<LockedDependency, ParseError> {
+        if line.contains("#egg=") {
+            let name = Self::parse_git_name(&line);
+            let git_ref = Self::parse_git_ref(&line)?;
```

Comment:
You don't need '&' here either!
You could write `Self::parse_git_ref(&&&&&&&line)` and that's valid code due to auto-deref, but it's really not needed =]